File: C++/DoubleLinkedList.cpp
#include <iostream>
#include <string>

class DoublyCircularLinkedList {
public:
    struct Node {
        std::string key;
        Node* prev;
        Node* next;

        Node(const std::string& k) : key(k), prev(nullptr), next(nullptr) {}
    };

    DoublyCircularLinkedList() {
        nil = new Node("NIL");
        nil->next = nil;
        nil->prev = nil;
    }

    ~DoublyCircularLinkedList() {
        Node* current = nil->next;
        while (current != nil) {
            Node* next = current->next;
            delete current;
            current = next;
        }
        delete nil;
    }

    void listInsert(Node* node) {
        node->next = nil->next;
        nil->next->prev = node;
        nil->next = node;
        node->prev = nil;
    }

    Node* listSearch(const std::string& key) {
        Node* x = nil->next;
        while (x != nil && x->key != key) {
            x = x->next;
        }
        return x;
    }

    void listDelete(Node* node) {
        if (node == nil) {
            std::cout << "Attempted to delete NIL node; skipping." << std::endl;
            return;
        }
        node->prev->next = node->next;
        node->next->prev = node->prev;
        delete node;
    }

    std::string display() {
        std::string result;
        Node* x = nil->next;
        while (x != nil) {
            result += x->key + " ";
            x = x->next;
        }
        if (!result.empty()) result.pop_back(); // 去掉最后一个空格
        return result;
    }

    Node* createNode(const std::string& key) {
        return new Node(key);
    }

private:
    Node* nil;
};

File: JAVA/DoubleLinkedList.java

    public class DoubleLinkedList {
        public DoubleLinkedList(Node nil) {
            this.nil = nil;
        }

        static class Node {
            String key;
            Node prev;
            Node next;

            Node(String key) {
                this.key = key;
                this.prev = this.next = null;
            }
        }

        private Node nil; // Sentinel node

        public void DoublyCircularLinkedList() {
            nil = new Node("NIL");
            nil.next = nil;
            nil.prev = nil;
        }

        public void listInsert(Node node) {
            // Insert at the head (right after NIL)
            node.next = nil.next;
            nil.next.prev = node;
            nil.next = node;
            node.prev = nil;
        }

        public Node listSearch(String key) {
            Node x = nil.next;
            while (x != nil && !x.key.equals(key)) {
                x = x.next;
            }
            return x;
        }

        public void listDelete(Node node) {
            if (node == nil) {
                System.out.println("Attempted to delete NIL node; skipping.");
                return;
            }
            node.prev.next = node.next;
            node.next.prev = node.prev;
        }

        public String display() {
            StringBuilder sb = new StringBuilder();
            Node x = nil.next;
            while (x != nil) {
                sb.append(x.key).append(" ");
                x = x.next;
            }
            return sb.toString().trim();
        }
    }


